### DISCUSSION TITLE
HSet() object, if hash not exists

### DISCUSSION THREAD

**Discussion by numberinho** [redis/rueidis#390] [Q&A]
Goal: Set a Hash with multiple fields, only if the hash does not already exist.

Why: I want to store a "flattened" object (containing only primitives as fields). Each field in the hash maps to a object property.

As far as I can see I cannot use `HSetNX`, because it only checks for the field to not already exist. Also, I cannot chain multiple `HSetNX`, because in concurrent access another routine may call `HGetAll` and not the whole object is stored in Redis yet. So `HGetAll` will miss some fields and return an incomplete object. Therefore I need some blocking mechanism.

Like a `HMSetNX` (this does not exist in Redis!), that would only set the fields if the *hash* does not exist.
This is a bit counter-intuitive, because one may interpret it to only set the fields if the *fields* do not exist.

For my understanding the function would look like:

```
lock
if hash not exists {
    HSet(object) -> HSetNX(key, value, key, value, ..)
}
lock
```

Can you give me any suggestions on how to do that using redis/rueidis? Maybe I need to chain the check for `exist` into a transaction with the `HSet` commands, to make sure nothing changed between `exist` and `HSet`?

Thank you!

**Comment by rueian** [comment]
Hi @numberinho, 

The most performant way is using a Lua script that first checks if `HLEN(key)` returns 0 and then does `HSET`, but you won't be able to use the `HSet` function of rueidiscompat.

**Comment by numberinho** [comment]
Hi @rueian 

That's a nice suggestion, thank you!

Not being able to use rediscompat is unfortunate. Rueidiscompat's `argsToSlice` in `adapter.go (line 3327)` would be  is super useful in this use case, but is not exported. On the other hand I think it would not make to much sense to export this function to keep consistency throughout the library.

**Reply by rueian** [reply]
Though making argToSlice and argsToSlice be public may let people confuse about how to use them, I think we can still make them public and give some usage examples.

**Comment by numberinho** [comment]
Hi @rueian I think I’m trying to reinvent the wheel…
Just got into the rueidis/om which provides what I want to achieve. Downside: It does not support all primitives and respective pointers , which I would love to add. What do you think?

Also, I can add some examples to argToSlice in the next days, if you want.

**Reply by numberinho** [reply]
Doing this, im pretty stuck at catching `time.Time`, which is a `reflect.struct`.

**Reply by rueian** [reply]
Hi @numberinho,

You could submit a draft PR first so that we can know how to help you.
